compiler/vmgen.nim

```diff
@@ -986,7 +986,14 @@ proc genMagic(c: PCtx; n: PNode; dest: var TDest; m: TMagic) =
     c.freeTemp(tmp)
     c.freeTemp(tmp2)
 
-  of mShlI: genBinaryABCnarrowU(c, n, dest, opcShlInt)
+  of mShlI:
+    genBinaryABC(c, n, dest, opcShlInt)
+    # genNarrowU modified
+    let t = skipTypes(n.typ, abstractVar-{tyTypeDesc})
+    if t.kind in {tyUInt8..tyUInt32} or (t.kind == tyUInt and t.size < 8):
+      c.gABC(n, opcNarrowU, dest, TRegister(t.size*8))
+    elif t.kind in {tyInt8..tyInt32} or (t.kind == tyInt and t.size < 8):
+      c.gABC(n, opcSignExtend, dest, TRegister(t.size*8))
```

Comment:
Are you sure you can't use `opcToNarrowInt` here? It performs the sign-extension but keeps it to the "interesting" part of the 64-bit register only.

Reply:
I also had a look at #10519 and using a similar trick fixes the test case provided there. Maybe `genNarrow` can be augmented with a `withRangeCheck = true` argument to select between `opcNarrowInt` (no check is performed) and `opcNarrowS`.

Maybe it's better to close the other PR and do all the fixes in this one.

Reply:
@LemonBoy I thought the same, but when I tried it the arguments didn't fit into a single opcode, so I introduce a new opcode, but then I also when I tried that I didn't just use the result register as the first argument, so maybe it could work. But I am not sure if it is worth the hassle since this opcode is pretty readable now. For #10519 I realized there is no narrowing necessary at all. Adding ``opcSignExtend`` there wouldn't do anything there. It would only make sense, if the input would be an illegal number (e.g. 1339 stored in an ``int8``), but I don't expect that at all since all arithmetic operations have the range check enabled (``opcNarrowS``).

Reply:
> I didn't just use the result register as the first argument, so maybe it could work

Yeah, if you specify `dst = src` the conversion is performed in-place.

> But I am not sure if it is worth the hassle since this opcode is pretty readable now

The amount of opcodes that can be encoded in a instruction is limited so adding a new one should not be done so lightly.

> Adding opcSignExtend there wouldn't do anything there

It does trim the register to the size of the type associated with it. Omitting it should not cause any issue but this means that `not 1'i8` now fills the whole 64-bits while all the other operations are careful to "clip" the results.

Reply:
> The amount of opcodes that can be encoded in a instruction is limited so adding a new one should not be done so lightly. 

You are right, but still I don't feel the need right now to change anything here. If your really care to join these two opcodes, please go ahead for a PR.  But right now there is still enough space for new opcodes.

> It does trim the register to the size of the type associated with it. Omitting it should not cause any issue but this means that `not 1'i8` now fills the whole 64-bits while all the other operations are careful to "clip" the results.

Well  to be precise ``not 1'i8`` would be computed in semfold, not here (yes that is redundent). But I know what you mean. And yes setting all bits to 1 in ``not 1'i8`` except the first one is exactly whas is supposed to happen. The sign extension would just redundantly ensure that all 56 bits on the left hand side of the sign bit are set to 1. Which is already the case.